### ISSUE TITLE
Bedrock module for Bake

### ISSUE THREAD

**Issue by carns** [mochi-hpc/mochi-bake#26]
In GitLab by @mdorier on Dec 3, 2020, 05:55

This issue lists the modifications to be done to create a Bedrock module for Bake.
* [x]  Follow the example [here](https://xgitlab.cels.anl.gov/sds/bedrock/-/blob/master/examples/module-a.c) and the header [here](https://xgitlab.cels.anl.gov/sds/bedrock/-/blob/master/include/bedrock/module.h) to implement a bake-bedrock.c file compiled into a libbake-bedrock.so library;
* [x]  Add the proper `--enable-bedrock` option in configure to look for the Bedrock header. There is no need to link libbake-bedrock.so with libbedrock-server.so, however it should be linked against libbake-server.so and libbake-client.so.
* [x]  Add a `+bedrock` option in Bake's spack package (I suggest turning it on by default) to enable building the Bedrock module.
* [x]  An ABT-IO instance is created in the file backend ([here](https://xgitlab.cels.anl.gov/sds/bake/-/blob/master/src/bake-file-backend.c#L184)). This instance should actually be owned by the provider, not the backend, and there should be a server-side function to set it, e.g. `int bake_provider_set_abt_io(bake_provider_t p, abt_io_instance_id abtioid)`. The Bedrock module should look get the optional ABT-IO instance out of the resolved dependencies and set it after registering the provider.
* [x]  Currently the Bake provider also instantiates a REMI client ([here](https://xgitlab.cels.anl.gov/sds/bake/-/blob/master/src/bake-server.c#L213)), and a REMI provider ([here](https://xgitlab.cels.anl.gov/sds/bake/-/blob/master/src/bake-server.c#L230)). These dependencies should be injected into the provider via an API rather than instantiated by the provider, i.e., there should be a function `bake_provider_set_remi(...)` in the bake server library, that takes a bake provider, a REMI provider and REMI client. (Matthieu can take care of that once the rest is done).
* [x]  We should be able to provide a JSON configuration when initializing the Bake provider. This can be done either by changing the signature of the `bake_provider_register` function (passing a structure like in Margo, containing the JSON string, the pool, and optional dependencies such as the ABT-IO instance and REMI client and provider. The JSON configuration should contain provider-specific configurations, as well as a "targets" array listing targets we want to open or create.
* [x]  We also need a bake provider to have a function to get a JSON object out such that passing this JSON object as config back to back will set Bake into exactly the same state as the one that generated the config (i.e. exact same parameters, exact same targets open, etc.).
* [x]  Margo now provides logging facility and Bedrock dependency-injects a logger in it, so if Bake were to use Margo's logging functions, its logs would integrate nicely with Bedrock's and Margo's. (see [here](https://xgitlab.cels.anl.gov/sds/margo/-/blob/master/include/margo-logging.h).
* [ ] It would be nice to have the admin library discussed [here](https://xgitlab.cels.anl.gov/sds/bake/-/issues/24) as well.
* [x] Something (either the module register or internal provider register) needs to auto create and/or auto attach targets that are specified ahead of time in the json
* [x] Update C++ bindings (after other changes)

**Comment by carns** [comment]
In GitLab by @carns on Jan 13, 2021, 18:30

Development is being done in https://xgitlab.cels.anl.gov/sds/bake/-/tree/dev-bedrock-26

**Comment by carns** [comment]
In GitLab by @carns on Jan 13, 2021, 18:30

marked the task **Add the proper `--enable-bedrock` option in configure to look for the Bedrock header. There is no need to link libbake-bedrock.so with libbedrock-server.so, however it should be linked against libbake-server.so and libbake-client.so.** as completed

**Comment by carns** [comment]
In GitLab by @carns on Jan 13, 2021, 18:30

marked the task **Add a `+bedrock` option in Bake's spack package (I suggest turning it on by default) to enable building the Bedrock module.** as completed

**Comment by carns** [comment]
In GitLab by @carns on Jan 26, 2021, 14:45

marked the task **We should be able to provide a JSON configuration when initializing the Bake provider. This can be done either by changing the signature of the `bake_provider_register` function (passing a structure like in Margo, containing the JSON string, the pool, and optional dependencies such as the ABT-IO instance and REMI client and provider. The JSON configuration should contain provider-specific configurations, as well as a "targets" array listing targets we want to open or create.** as completed

**Comment by carns** [comment]
In GitLab by @carns on Jan 26, 2021, 14:45

marked the task **We also need a bake provider to have a function to get a JSON object out such that passing this JSON object as config back to back will set Bake into exactly the same state as the one that generated the config (i.e. exact same parameters, exact same targets open, etc.).** as completed

**Comment by carns** [comment]
In GitLab by @carns on Jan 26, 2021, 15:10

changed the description

**Comment by carns** [comment]
In GitLab by @carns on Jan 27, 2021, 16:03

marked the task **Follow the example [here](https://xgitlab.cels.anl.gov/sds/bedrock/-/blob/master/examples/module-a.c) and the header [here](https://xgitlab.cels.anl.gov/sds/bedrock/-/blob/master/include/bedrock/module.h) to implement a bake-bedrock.c file compiled into a libbake-bedrock.so library;** as completed

**Comment by carns** [comment]
In GitLab by @carns on Jan 27, 2021, 17:11

changed the description

**Comment by carns** [comment]
In GitLab by @carns on Jan 27, 2021, 17:37

marked the task **Margo now provides logging facility and Bedrock dependency-injects a logger in it, so if Bake were to use Margo's logging functions, its logs would integrate nicely with Bedrock's and Margo's. (see [here](https://xgitlab.cels.anl.gov/sds/margo/-/blob/master/include/margo-logging.h).** as completed

**Comment by carns** [comment]
In GitLab by @carns on Feb 1, 2021, 08:42

marked the task **Something (either the module register or internal provider register) needs to auto create and/or auto attach targets that are specified ahead of time in the json** as completed

**Comment by carns** [comment]
In GitLab by @carns on Feb 1, 2021, 08:43

mentioned in merge request !19

**Comment by carns** [comment]
In GitLab by @carns on Feb 15, 2021, 18:10

marked the task **An ABT-IO instance is created in the file backend ([here](https://xgitlab.cels.anl.gov/sds/bake/-/blob/master/src/bake-file-backend.c#L184)). This instance should actually be owned by the provider, not the backend, and there should be a server-side function to set it, e.g. `int bake_provider_set_abt_io(bake_provider_t p, abt_io_instance_id abtioid)`. The Bedrock module should look get the optional ABT-IO instance out of the resolved dependencies and set it after registering the provider.** as completed

**Comment by carns** [comment]
In GitLab by @carns on Feb 15, 2021, 18:11

I propose deferring admin library to after this is merged (it's already tracked in a separate issue).

That just leaves c++ updates and remi updates.

**Comment by carns** [comment]
In GitLab by @mdorier on Feb 17, 2021, 05:26

marked the task **Currently the Bake provider also instantiates a REMI client ([here](https://xgitlab.cels.anl.gov/sds/bake/-/blob/master/src/bake-server.c#L213)), and a REMI provider ([here](https://xgitlab.cels.anl.gov/sds/bake/-/blob/master/src/bake-server.c#L230)). These dependencies should be injected into the provider via an API rather than instantiated by the provider, i.e., there should be a function `bake_provider_set_remi(...)` in the bake server library, that takes a bake provider, a REMI provider and REMI client. (Matthieu can take care of that once the rest is done).** as completed

**Comment by carns** [comment]
In GitLab by @mdorier on Feb 17, 2021, 05:26

marked the task **Update C++ bindings (after other changes)** as completed

**Comment by carns** [comment]
In GitLab by @carns on Feb 18, 2021, 07:28

This is merged to master now.  Closing ticket and tracking admin lib separately.

Some codes that use bake need to be update but that task can be tracked separately as well.

**Comment by carns** [comment]
In GitLab by @carns on Feb 18, 2021, 07:28

closed
